Move CatBoost evaluation into clickhouse-library-bridge

rschu1ze opened the pull request:
This PR re-introduces #39629 which has been reverted (#40877).

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
For security and stability reasons, catboost models are no longer evaluated within the ClickHouse server. Instead, the evaluation is now done in the clickhouse-library-bridge, a separate process that loads the catboost library and communicates with the server process via HTTP.

#### Upgrade notes

Function `modelEvaluate()` function is no longer available. To evaluate catboost models, user can now use the new function `catboostEvaluate()`, documented [here](https://clickhouse.com/docs/en/sql-reference/functions/other-functions/#catboostevaluatepath_to_model-feature_1-feature_2--feature_n):

```
  SELECT
    FEAT_1, ..., FEAT_N, catboostEvaluate('/path/to/model.bin', FEAT_1, ..., FEAT_N)
  FROM data
  LIMIT 10
```
The syntax of function`catboostEvaluate()` is similar to the previous `modelEvaluate()`, but instead of specifying a model name which is mapped to a model path using an XML file, users now specify the model path directly.

Required configuration:
```
  <catboost_lib_path>/path/to/libcatboostmodel.so</catboost_lib_path>
```
Configuration parameter `catboost_lib_path` corresponds to the previous configuration parameter `catboost_dynamic_library_path` which is now ignored.

Fixes #27870

rschu1ze committed:
feat: implement catboost in library-bridge

This commit moves the catboost model evaluation out of the server
process into the library-bridge binary. This serves two goals: On the
one hand, crashes / memory corruptions of the catboost library no longer
affect the server. On the other hand, we can forbid loading dynamic
libraries in the server (catboost was the last consumer of this
functionality), thus improving security.

SQL syntax:

  SELECT
    catboostEvaluate('/path/to/model.bin', FEAT_1, ..., FEAT_N) > 0 AS prediction,
    ACTION AS target
  FROM amazon_train
  LIMIT 10

Required configuration:

  <catboost_lib_path>/path/to/libcatboostmodel.so</catboost_lib_path>

*** Implementation Details ***

The internal protocol between the server and the library-bridge is
simple:

- HTTP GET on path "/extdict_ping":
  A ping, used during the handshake to check if the library-bridge runs.

- HTTP POST on path "extdict_request"
  (1) Send a "catboost_GetTreeCount" request from the server to the
      bridge, containing a library path (e.g /home/user/libcatboost.so) and
      a model path (e.g. /home/user/model.bin). Rirst, this unloads the
      catboost library handler associated to the model path (if it was
      loaded), then loads the catboost library handler associated to the
      model path, then executes GetTreeCount() on the library handler and
      finally sends the result back to the server. Step (1) is called once
      by the server from FunctionCatBoostEvaluate::getReturnTypeImpl(). The
      library path handler is unloaded in the beginning because it contains
      state which may no longer be valid if the user runs
      catboost("/path/to/model.bin", ...) more than once and if "model.bin"
      was updated in between.
  (2) Send "catboost_Evaluate" from the server to the bridge, containing
      the model path and the features to run the interference on. Step (2)
      is called multiple times (once per chunk) by the server from function
      FunctionCatBoostEvaluate::executeImpl(). The library handler for the
      given model path is expected to be already loaded by Step (1).

Fixes #27870

rschu1ze committed:
chore: delete obsolete modelEvaluate() function + SYSTEM.MODELS view

- The deleted function modelEvaluate() was superseded by
  catboostEvaluate().

- Also delete the external model repository, as modelEvaluate() was it's
  last user. Additionally remove the system view SYSTEM.MODELS for
  inspecting the repository.

- SYSTEM RELOAD MODELS is also obsolete. HOWEVER, it was retained and
  made a no-op instead of deleted.

  Why?
  The reason is that RBAC in distributed setups works by storing
  privileges (granted and revoked) as plain SQL statements in Keeper.
  Nodes read these statements at startup and parse them. If a privilege
  for SYSTEM RELOAD MODELS exists but parser doesn't recognize it
  nodes would fail to come up.

  Considered but rejected alternatives:
  - Ignore SYSTEM RELOAD MODELS during parsing RBAC privileges and
    return an error for regular SYSTEM RELOAD MODELS SQL. Special-case
    of no-op behavior, too brittle.
  - Remove SYSTEM RELOAD MODELS manually from Keeper via command-line
    manipulation of Keeper nodes or via SQL by dropping the privileges.
    Needs user intervention during upgrade.

rschu1ze commented on src/Interpreters/InterpreterSystemQuery.cpp:
```diff
@@ -374,19 +373,10 @@ BlockIO InterpreterSystemQuery::execute()
-
-            auto & external_models_loader = system_context->getExternalModelsLoader();
-            external_models_loader.reloadAllTriedToLoad();
+            /// SYSTEM RELOAD MODEL(S) is no longer supported. For compat reasons, it is not completely removed but retained as no-op.
```
Fix for RBAC issue.

rschu1ze replied:
I spend quite some time today to verify locally that the issue found by Nikita in the cloud (aka. reason for revert #40877) no longer happens.

Overall, I think "normal" users would not have run into the issue anyway - 1. it requires granting the [*undocumented*](https://clickhouse.com/docs/en/sql-reference/statements/grant/#grant-select) `SYSTEM RELOAD MODELS` privilege to a user/role and 2. ... which is needed only for the slightly obscure catboost usecase. My guess is that Nikitas cloud machine configuration has some history and is pretty "wild".

(Just FYI for the record: Besides `SYSTEM RELOAD MODELS`, this PR also removes view `SYSTEM.MODELS`. In theory, there could be a similar problem for SELECT privileges on this view. In practice, there isn't because SELECT privileges can be [defined on non-existing tables](https://clickhouse.com/docs/en/operations/access-rights/#access-control-properties).)

To sum up, this PR is good to merge in my POV.

UnamedRus commented:
> SYSTEM RELOAD MODEL(S) is no longer supported

That is really sad, as we know multiple companies which used this feature.

rschu1ze commented:
The catboost feature itself is retained but in a way that `SYSTEM RELOAD MODELS` is no longer needed. In fact, due to a last-minute bug, `SYSTEM RELOAD MODELS` continues to exist but as a no-op (see https://github.com/ClickHouse/ClickHouse/pull/40897/commits/0183a370414a400573fd70d8f4d71a7a81a90670 for details)

UnamedRus commented:
Yes, i understand that catboost models now works via http bridge like jdbc/odbc, but it's unclear for me for example, what will happen if we will replace model .bin file? 
Will bridge auto-reload model by itself?
There is another concern about performance penalty, but i guess it inevitable for now.

rschu1ze commented:
Each call to `catboostEvaluate()` loads the model .bin freshly from disk (the path to the .bin file must now be specified in SQL). So updates are supported. The cost of reloading can be amortized by evaluating multiple rows at once. The maximum HTTP URL length however limits the number of rows to be evaluated at once, in my tests the threshold was around 50k rows (but that depends on the number of features).

Yes, HTTP communication with the library-bridge has a performance cost but that's inevitable ...

kssenii commented:
> Each call to catboostEvaluate() loads the model .bin freshly from disk

btw why not reuse the loaded result and reload it only when explicitly said to do that? For example, library bridge does not load the library for each request and reloads it only when the library dictionary source is reloaded.

rschu1ze commented:
It works as designed :)
To be serious, there were 3 reasons for reloading the model with each call: 1. My assumption was that catboost evaluation is a quite exotic feature and only used for experiments/demonstration, but that's obviously not true as per https://github.com/ClickHouse/ClickHouse/pull/40897#issuecomment-1237405340. My aim was to keep the logic after migration to library-bridge as simple as possible 2. The .bin files contain a bunch of integers as decision tree parameters and are not too big (in average). I worked only with two models, they were 0.8 MB and 4.8 MB in size. 3. The cost of loading the model is amortized by doing so only once per batch evaluation (e.g. 50k rows).

@kssenii Will change to the behavior you described (shouldn't be more than a few hours work). This will also resurrect `SYSTEM RELOAD MODELS`. (I thought about merging this PR and doing the fix separately but keeping it around for a bit longer doesn't matter too much)

rschu1ze committed:
chore: restore SYSTEM RELOAD MODEL(S) and moniting view SYSTEM.MODELS

- This commit restores statements "SYSTEM RELOAD MODEL(S)" which provide
  a mechanism to update a model explicitly. It also saves potentially
  unnecessary reloads of a model from disk after it's initial load.

  To keep the complexity low, the semantics of "SYSTEM RELOAD MODEL(S)
  was changed from eager to lazy. This means that both statements
  previously immedately reloaded the specified/all models, whereas now
  the statements only trigger an unload and the first call to
  catboostEvaluate() does the actual load.

- Monitoring view SYSTEM.MODELS is also restored but with some obsolete
  fields removed. The view was not documented in the past and for now it
  remains undocumented. The commit is thus not considered a breach of
  ClickHouse's public interface.

rschu1ze commented:
I restored `SYSTEM RELOAD MODEL(S)` and also system view `SYSTEM.MODELS`. There are still minor caveats where the new behavior is slightly different from the previous behavior (documented in [this commit msg](https://github.com/ClickHouse/ClickHouse/pull/40897/commits/484c541acd07a035a5972d42de9f6f482720b944)). But I think that should be okay.

rschu1ze committed:
fix: skip SYSTEM.MODELS In test 01161_all_system_tables.sh

The test is a stress test which queries all system tables in parallel.
System table SYSTEM.MODELS is now populated using the library-bridge.
This made the test fail with error:

   2022-09-12 23:51:34 Code: 410. DB::Exception: Received from localhost:9000. DB::Exception: BridgeHelper: clickhouse-library-bridge is not responding. (EXTERNAL_SERVER_IS_NOT_RESPONDING)
   2022-09-12 23:51:34 (query: SELECT * FROM system.models LIMIT 10000 FORMAT Null)

Looking at the logs, the server tried to start the library-bridge when
querying SYSTEM.MODELS but multiple handshake attempts failed.

So (most likely) the infrastructure is slow or (unlikely) there are
settings in CI (e.g. a firewall) preventing communication with the
bridge. Because 1. other system tables are also excluded (zookeeper,
merge_tree_metadata_cache), 2. SYSTEM.MODELS is tested in integration
test "test_catboost_evaluate" and 3. the test runs locally just fine, I
am excluding SYSTEM.MODELS from it.

rschu1ze committed:
fix: repair broken documentation links and update docs

rschu1ze commented:
@kssenii This PR is done IMHO and the tests for the last run were also fine. Green light would be great. Thanks!

kssenii commented on src/BridgeHelper/CatBoostLibraryBridgeHelper.h:
```diff
@@ -17,9 +18,15 @@ class CatBoostLibraryBridgeHelper : public LibraryBridgeHelper
+    ExternalModelInfos listModels();
+
+    void removeModel();
+    void removeAllModels();
```
this seems a little inconsistent that model_path can be an argument in bridge helpers constructor and at the same not be, e.g. some methods need to ensure that bridge helper was created with a certain constructor. May be rework it somehow to avoid such complication?

kssenii commented on src/Interpreters/InterpreterSystemQuery.cpp:
```diff
@@ -373,10 +374,17 @@ BlockIO InterpreterSystemQuery::execute()
+        {
+            getContext()->checkAccess(AccessType::SYSTEM_RELOAD_MODEL);
+            auto bridge_helper = std::make_unique<CatBoostLibraryBridgeHelper>(getContext(), query.target_model);
+            bridge_helper->removeModel();
```
why not actually reload and not just remove? I would expect that time for reloading the model lies on the reload model command instead of the next attempt to access it

rschu1ze replied:
I agree in general (even if the reload costs are small because models are small)

The problem is that the model evaluation is not an atomic operation. It is actually two operations A and B with a dependency between them. This cannot be changed easily because of ClickHouse's function interface.

So we need to deal with a `SYSTEM RELOAD MODEL` call between A and B. The simplest approach is the one implemented: Simply unload the model, and let B check if the model is still in memory. If A finds a model unloaded, load it.

If `SYSTEM RELOAD MODEL` actually *reloads* the model, then B must somehow find that out and fail. There are multiple approaches. I'll mention just two
- A locks the model on the bridge side until B starts. This avoids reloads in between. Obviously there is a risk of livelocks here ...
- More sophisticated: Generated for each `catboostEvaluate()` call a random "request id" on the server side, send it to the bridge in A and store it in a queue. A reload in between clears the queue. The same request id is send to the bridge in B. If it is not found in the queue, fail. The length of the queue limits the number of parallel requests which is why I did not like this option too much. It's also harder to maintain and test.

TLDL: I think the current behavior is not a deal breaker for such quite exotic functionality. If it's okay, I'll ask Nikolai what he thinks (he was the original author) and if needed implement an actual load in a separate PR.

rschu1ze replied:
For completeness, I talked briefly to Nikolai yesterday and he's fine with the lazy semantics.

rschu1ze committed:
fix: incorporate review comments
